### ISSUE TITLE
displaylink / evdi build broken

### ISSUE THREAD

**Issue by umeat** [NixOS/nixpkgs#200124]
### Steps To Reproduce

`services.xserver.videoDrivers = [ "modesetting" "displaylink" ];`

### Build log
```
...
make[3]: *** [/nix/store/9si6d01y28kzj3g3gbpfdx4wsk66v8rq-linux-6.0.6-dev/lib/modules/6.0.6/source/scripts/Makefile.build:249: /tmp/nix-build-evdi-1.12.0.drv-0/source/module/evdi_painter.o] Error 1
make[2]: *** [/nix/store/9si6d01y28kzj3g3gbpfdx4wsk66v8rq-linux-6.0.6-dev/lib/modules/6.0.6/source/Makefile:1852: /tmp/nix-build-evdi-1.12.0.drv-0/source/module] Error 2
make[2]: Leaving directory '/nix/store/9si6d01y28kzj3g3gbpfdx4wsk66v8rq-linux-6.0.6-dev/lib/modules/6.0.6/build'
make[1]: *** [Makefile:81: module] Error 2
make[1]: Leaving directory '/tmp/nix-build-evdi-1.12.0.drv-0/source/module'
make: *** [Makefile:8: all] Error 2
error: builder for '/nix/store/nlrk3sf2yjdns2xs71z5nf5c7yfj253z-evdi-1.12.0.drv' failed with exit code 2
error: 1 dependencies of derivation '/nix/store/wda9jvbsxhya04328klywfbvsqiswi0z-displaylink-5.6.1-59.184.drv' failed to build
error: 1 dependencies of derivation '/nix/store/h8ild7yyr32lkgxvlw21gq11710lq9y2-kernel-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/9vcqxgl86bvl6djqvgpz12lgr72vs84a-linux-6.0.6-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/lw459505n3q1pkix438bhb6z1zgw5ars-nixos-system-nosleep-22.11pre423979.3bacde6273b.drv' failed to build
```

### Notify maintainers

@abbradar
@peterhoeg
@eyJhb

### Metadata

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.0.6, NixOS, 22.11 (Raccoon), 22.11pre423979.3bacde6273b`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by rhoriguchi** [comment]
Broken since release d40fea9aeb8840fea0d377baa4b38e39b9582458.

**Comment by ProducerMatt** [comment]
I'm encountering fails as well. My error log is very long but here's the tail end:

```
/build/source/module/evdi_painter.c: In function 'evdi_painter_grabpix_ioctl':
/build/source/module/evdi_painter.c:1060:9: error: implicit declaration of function 'drm_framebuffer_get' [-Werror=implicit-function-declaration]
 1060 |         drm_framebuffer_get(&efb->base);
      |         ^~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[3]: *** [/nix/store/gs8ciyzpr817jpf8b2bp3s3dgkp3x088-linux-6.1.8-dev/lib/modules/6.1.8/source/scripts/Makefile.build:250: /build/source/module/evdi_painter.o] Error 1
/build/source/module/evdi_fb.c: In function 'evdi_fbdev_destroy':
/build/source/module/evdi_fb.c:472:17: error: implicit declaration of function 'drm_framebuffer_unregister_private' [-Werror=implicit-function-declaration]
  472 |                 drm_framebuffer_unregister_private(&efbdev->efb.base);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/source/module/evdi_fb.c: At top level:
/build/source/module/evdi_fb.c:331:43: error: storage size of 'evdifb_funcs' isn't known
  331 | static const struct drm_framebuffer_funcs evdifb_funcs = {
      |                                           ^~~~~~~~~~~~
/build/source/module/evdi_fb.c: In function 'evdi_framebuffer_init':
/build/source/module/evdi_fb.c:350:1: error: control reaches end of non-void function [-Werror=return-type]
  350 | }
      | ^
cc1: some warnings being treated as errors
make[3]: *** [/nix/store/gs8ciyzpr817jpf8b2bp3s3dgkp3x088-linux-6.1.8-dev/lib/modules/6.1.8/source/scripts/Makefile.build:250: /build/source/module/evdi_fb.o] Error 1
make[2]: *** [/nix/store/gs8ciyzpr817jpf8b2bp3s3dgkp3x088-linux-6.1.8-dev/lib/modules/6.1.8/source/Makefile:1992: /build/source/module] Error 2
make[2]: Leaving directory '/nix/store/gs8ciyzpr817jpf8b2bp3s3dgkp3x088-linux-6.1.8-dev/lib/modules/6.1.8/build'
make[1]: *** [Makefile:81: module] Error 2
make[1]: Leaving directory '/build/source/module'
make: *** [Makefile:8: all] Error 2
```

**Comment by ProducerMatt** [comment]
Workaround found with the following overlay:

```nix
final: prev: {
  linuxPackages_latest =
    prev.linuxPackages_latest.extend
      (lpfinal: lpprev: {
        evdi =
          lpprev.evdi.overrideAttrs (efinal: eprev: {
            version = "1.12.0-git";
            src = prev.fetchFromGitHub {
              owner = "DisplayLink";
              repo = "evdi";
              rev = "bdc258b25df4d00f222fde0e3c5003bf88ef17b5";
              sha256 = "sha256-mt+vEp9FFf7smmE2PzuH/3EYl7h89RBN1zTVvv2qJ/o=";
            };
          });
      });
  displaylink = prev.displaylink.override {
    inherit (final.linuxPackages_latest) evdi;
  };
}
```

At least it builds. I can't test it because I can't get the display to show up, but that's likely unrelated to the issue.

**Comment by rhoriguchi** [comment]
This is fixed with 7f5639fa3b68054ca0b062866dc62b22c3f11505, might work with earlier commits.

**Comment by segator** [comment]
I tried to apply the patch mentioned in kernel 6.7.2 and it fails

```
/nix/store/zrqk5y7p1dk040a22cca8cy3hm978x9a-linux-6.7.2-dev/lib/modules/6.7.2/source/include/linux/dma-mapping.h:89:>
   89 |                 unsigned long len)
      |                 ~~~~~~~~~~~~~~^~~
/build/source/module/evdi_painter.c: At top level:
/build/source/module/evdi_painter.c:1308:6: error: conflicting types for 'evdi_painter_i2c_data_notify'; have 'bool(>
 1308 | bool evdi_painter_i2c_data_notify(struct evdi_painter *painter, struct i2c_msg *msg)
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/source/module/evdi_drm_drv.h:196:6: note: previous declaration of 'evdi_painter_i2c_data_notify' with type 'b>
  196 | bool evdi_painter_i2c_data_notify(struct evdi_painter *painter, struct i2c_msg *msg);
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
